homeassistant/components/climate/generic_thermostat.py

```diff
@@ -302,65 +390,143 @@ def _async_update_temp(self, state):
+                if not long_enough_heating:
+                    return
+            if self.current_operation == STATE_AUTO:
+                if not long_enough_cooling or not long_enough_heating:
+                    return
+
+        if self.current_operation == STATE_COOL:
+            if self._dual_temp:
+                target_temp = self._target_temp_high
             else:
-                too_hot = self._cur_temp - self._target_temp >= \
-                    self._hot_tolerance
-                if too_hot:
-                    _LOGGER.info('Turning on AC %s', self.heater_entity_id)
-                    self._heater_turn_on()
-        else:
-            is_heating = self._is_device_active
-            if is_heating:
-                too_hot = self._cur_temp - self._target_temp >= \
-                    self._hot_tolerance
-                if too_hot:
-                    _LOGGER.info('Turning off heater %s',
-                                 self.heater_entity_id)
-                    self._heater_turn_off()
+                target_temp = self._target_temp
+            if self.ac_entity_id:
+                if self._is_cooling:
+                    too_cold = target_temp - self._cur_temp >= \
+                        self._cold_tolerance
+                    if too_cold:
+                        _LOGGER.info('Turning off AC %s', self.ac_entity_id)
+                        self._ac_turn_off()
+                else:
+                    too_hot = self._cur_temp - target_temp >= \
+                        self._hot_tolerance
+                    if too_hot:
+                        _LOGGER.info('Turning on AC %s', self.ac_entity_id)
+                        self._ac_turn_on()
+
+        if self.current_operation == STATE_HEAT:
+            if self._dual_temp:
+                target_temp = self._target_temp_low
             else:
-                too_cold = self._target_temp - self._cur_temp >= \
-                    self._cold_tolerance
-                if too_cold:
-                    _LOGGER.info('Turning on heater %s', self.heater_entity_id)
-                    self._heater_turn_on()
+                target_temp = self._target_temp
+            if self.heater_entity_id:
+                if self._is_heating:
+                    too_hot = self._cur_temp - target_temp >= \
+                        self._hot_tolerance
+                    if too_hot:
+                        _LOGGER.info('Turning off heater %s',
+                                     self.heater_entity_id)
+                        self._heater_turn_off()
+                else:
+                    too_cold = target_temp - self._cur_temp >= \
+                        self._cold_tolerance
+                    if too_cold:
+                        _LOGGER.info('Turning on heater %s',
+                                     self.heater_entity_id)
+                        self._heater_turn_on()
+
+        if self.current_operation == STATE_AUTO:
```

Comment:
As said above, I think you can completely get rid of this block.

Reply:
Removed more blocks actually with your suggestions